Lib/test/test_concurrent_futures.py

```diff
@@ -134,6 +143,23 @@ def _prime_executor(self):
         for f in futures:
             f.result()
 
+    def _fail_on_deadlock(self, executor=None):
```

Comment:
Is this useful? You could instead pass `--timeout` to `python -m test`.

Reply:
`--timeout` does not work with the `-j` option so it lead to some deadlocks in the CI.
This option is safer as it will always work.

Reply:
I don't want to add test-specific code to workaround issues in our test runner. Perhaps @vstinner wants to take a look why `--timeout` doesn't work with `-j`, but I don't think this PR should be complicated by this.

Reply:
After some investigation:

- The `--timeout` does work with `-j` but the dumped tracebacks are non informative: they qre tb of the runner and it happens before the indiviual test subprocesses have a chance to timeout by themselves. We did change that in 65f3aba so the timeout is now per test only.

- However, this does not fix all the issues. The `faulthandler.dump_traceback_later` with `exit=True` hang when the exited test process also has subprocesses by itself. In our test, the subprocesses will never terminate in case of a deadlock, as the `queue_manager_thread` does not exist anymore at this point. The solution would be to make `faulthandler` recursively terminate all subprocesses when exiting. There is a related open issue for `faulthandler` : https://bugs.python.org/issue12413

We remove the test-specific timers but we kept the timeout for `test_crashes` in order to avoid stalling CI for 1hour+ with tests that try to trigger deadlocks.